v2/reader.go

```diff
@@ -121,3 +124,62 @@ func ReadVersion(r io.Reader) (version uint64, err error) {
 	}
 	return header.Version, nil
 }
+
+var _ io.ReaderAt = (*readSeekerAt)(nil)
+
+type readSeekerAt struct {
+	rs io.ReadSeeker
+	mu sync.Mutex
+}
+
+func (rsa *readSeekerAt) ReadAt(p []byte, off int64) (n int, err error) {
+	rsa.mu.Lock()
+	defer rsa.mu.Unlock()
+	if _, err := rsa.rs.Seek(off, io.SeekStart); err != nil {
+		return 0, err
+	}
+	return rsa.rs.Read(p)
+}
+
+// ReadOrGenerateIndex accepts both CAR v1 and v2 format, and reads or generates an index for it.
+// When the given reader is in CAR v1 format an index is always generated.
+// For a payload in CAR v2 format, an index is only generated if Header.HasIndex returns false.
+// An error is returned for all other formats, i.e. versions other than 1 or 2.
+//
+// Note, the returned index lives entirely in memory and will not depend on the
+// given reader to fulfill index lookup.
+func ReadOrGenerateIndex(rs io.ReadSeeker) (index.Index, error) {
```

Comment:
there should be a way to do this on just an `io.Reader` without ever needing to seek, but we would end up needing to use internal functions to jump to the internal (post-header) portion of index generation when we encounter a car v1 header.

Reply:
I agree; this should be doable with an `io.Reader` alone. `Seek` is mainly used for skipping over blocks when generating index, otherwise we would have to read and discard while iterating through blocks during indexing. 

Would it make sense to capture an issue and mark it as a P3 deferring implementation unless folks explicitly ask for it?

Reply:
yeah. that works. I'll leave relative priority to you. I suspect in practice it doesn't matter _too_ much, but i can imagine situations where you're streaming a car onto disk, and want to generate an index while you're at it. Being able to pipe through this with just an `io.Reader` is likely the way you'd hope to do that, so we should eventually relax the interface here to what we actually need.

Reply:
https://github.com/ipld/go-car/issues/146